### PR TITLE
Includes/Autoloader: Change - Make auto loader extend able. ED-1065

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: includes/autoloader.php [hunk 1]
```diff
@@ -129,6 +157,10 @@ private static function init_classes_map() {
 			'WordPress_Widgets_Manager' => 'includes/managers/wordpress-widgets.php',
 		];
 
+		if ( ! class_exists(  Controls_Manager::class ) ) {
```
Comment (matipojo) [inline includes/autoloader.php:160]: Why?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/autoloader.php (hunk 1 above)
Comment (iNewLegend) [reply includes/autoloader.php]: Handled.

File: includes/autoloader.php [hunk 2]
```diff
@@ -129,6 +157,11 @@ private static function init_classes_map() {
 			'WordPress_Widgets_Manager' => 'includes/managers/wordpress-widgets.php',
 		];
 
+		// Who ever uses autoloader, should load 'Controls_Manager' before to support it functionality.
```
Comment (matipojo) [inline includes/autoloader.php:160]: Extract a base autoloader with an empty `init_classes_map`, extend it from the core and move current `init_classes_map` to the new extended class.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/autoloader.php (hunk 2 above)
Comment (iNewLegend) [reply includes/autoloader.php]: Handled.

File: tests/bootstrap.php [hunk 3]
```diff
@@ -42,7 +42,7 @@
 require __DIR__ . '/phpunit/factories/factory.php';
 require __DIR__ . '/phpunit/factories/documents.php';
 
-require_once dirname( __DIR__ ) . '/includes/autoloader.php';
+require_once dirname( __DIR__ ) . '/core/autoloader.php';
```
Comment (matipojo) [inline tests/bootstrap.php:45]: Restore the file path for BC.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/bootstrap.php (hunk 3 above)
Comment (iNewLegend) [reply tests/bootstrap.php]: Handled.

File: includes/autoloader.php [hunk 4]
```diff
@@ -1,314 +1,154 @@
...
-	 *
-	 * @since 1.6.0
-	 * @access public
-	 * @static
-	 *
-	 * @return array Classes aliases.
-	 */
+class Autoloader extends Autoloader_Base {
 	public static function get_classes_aliases() {
 		if ( ! self::$classes_aliases ) {
```
Comment (matipojo) [inline includes/autoloader.php:8]: Use `static` instead of self - in all places.

File: includes/plugin.php [hunk 5]
```diff
@@ -2,6 +2,7 @@
 namespace Elementor;
 
 use Elementor\Core\Admin\Admin;
+use Elementor\Autoloader as AutoLoader;
```
Comment (matipojo) [inline includes/plugin.php:5]: Why?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/autoloader.php (hunk 4 above)
Comment (iNewLegend) [reply includes/autoloader.php]: Handled.

File: includes/plugin.php (hunk 5 above)
Comment (iNewLegend) [reply includes/plugin.php]: Handled.

File: includes/plugin.php [hunk 6]
```diff
@@ -691,7 +692,7 @@ private function add_cpt_support() {
 	private function register_autoloader() {
 		require_once ELEMENTOR_PATH . '/includes/autoloader.php';
 
-		Autoloader::run();
+		AutoLoader::run();
```
Comment (matipojo) [inline includes/plugin.php:695]: Why?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/plugin.php (hunk 6 above)
Comment (iNewLegend) [reply includes/plugin.php]: Handled.

**Review by matipojo — DISMISSED** [review]

File: includes/base/autoloader-base.php [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	 *
+	 * Maps Elementor classes to file names.
+	 *
+	 * @since 1.6.0
+	 * @access private
+	 * @static
+	 *
+	 * @var array Classes used by elementor.
+	 */
+	protected static $classes_map;
```
Comment (kobizz) [inline includes/base/autoloader-base.php:29]: Move classes map to `includes/autoloader.php`

File: includes/autoloader.php [hunk 8]
```diff
@@ -1,314 +1,154 @@
...
-	 * @access public
-	 * @static
-	 *
-	 * @return array Classes aliases.
-	 */
+class Autoloader extends Autoloader_Base {
 	public static function get_classes_aliases() {
-		if ( ! self::$classes_aliases ) {
-			self::init_classes_aliases();
+		if ( ! static::$classes_aliases ) {
```
Comment (kobizz) [inline includes/autoloader.php:8]: The init of `classes_aliases` should be done in the base and stay private. The extending class should only provide a list of its own aliases

**Review by kobizz — CHANGES_REQUESTED** [review]

File: includes/base/autoloader-base.php (hunk 7 above)
Comment (iNewLegend) [reply includes/base/autoloader-base.php]: Handled.

File: includes/autoloader.php (hunk 8 above)
Comment (iNewLegend) [reply includes/autoloader.php]: Handled.

**Review by matipojo — CHANGES_REQUESTED** [review]
Re-extract a clean base class without the aliases functionality.

File: includes/base/autoloader-base.php [hunk 9]
```diff
@@ -0,0 +1,278 @@
+<?php
```
Comment (matipojo) [inline includes/base/autoloader-base.php:1]: Move all `classes_map` related code to the `Autoloader`.
the base class shouldn't know about it.

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/base/autoloader-base.php (hunk 9 above)
Comment (iNewLegend) [reply includes/base/autoloader-base.php]: No require base, and useless. look pm for more info.

**Review by matipojo — APPROVED** [review]

**Review by kobizz — APPROVED** [review]

**Pull request merged**